### a)

In [1]:
import pandas as pd
df=pd.read_csv("C:/Users/iarak/OneDrive/Desktop/persp-model-econ_W20/ProblemSets/PS6/data/Auto.csv", na_values='?')

### b)

In [2]:
from pandas.plotting import scatter_matrix
import matplotlib.pyplot as plt
plt.style.use('ggplot')
df.dropna(inplace=True)
df_quant=df[['mpg','cylinders','displacement','horsepower',
             'weight','acceleration','year','origin']]
df_quant=df_quant.astype(float)
scatter_matrix(df_quant, alpha=0.3,figsize=(6,6),diagonal='kde')
plt.show()

<Figure size 600x600 with 64 Axes>

### c) Correlation Matrix

In [3]:
df_quant.corr()

mpg  cylinders  displacement  horsepower    weight  \
mpg           1.000000  -0.777618     -0.805127   -0.778427 -0.832244   
cylinders    -0.777618   1.000000      0.950823    0.842983  0.897527   
displacement -0.805127   0.950823      1.000000    0.897257  0.932994   
horsepower   -0.778427   0.842983      0.897257    1.000000  0.864538   
weight       -0.832244   0.897527      0.932994    0.864538  1.000000   
acceleration  0.423329  -0.504683     -0.543800   -0.689196 -0.416839   
year          0.580541  -0.345647     -0.369855   -0.416361 -0.309120   
origin        0.565209  -0.568932     -0.614535   -0.455171 -0.585005   

              acceleration      year    origin  
mpg               0.423329  0.580541  0.565209  
cylinders        -0.504683 -0.345647 -0.568932  
displacement     -0.543800 -0.369855 -0.614535  
horsepower       -0.689196 -0.416361 -0.455171  
weight           -0.416839 -0.309120 -0.585005  
acceleration      1.000000  0.290316  0.212746  
year              0.290316  1.000000  0.181528  
origin            0.212746  0.181528  1.000000

### d)

In [4]:
import statsmodels.api as sm
y = df_quant['mpg']
x = sm.add_constant(df_quant[['cylinders','displacement','horsepower','weight',
                              'acceleration','year','origin']])
reg1 = sm.OLS(y, x)
results = reg1.fit()
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.821
Model:                            OLS   Adj. R-squared:                  0.818
Method:                 Least Squares   F-statistic:                     252.4
Date:                Tue, 18 Feb 2020   Prob (F-statistic):          2.04e-139
Time:                        20:39:44   Log-Likelihood:                -1023.5
No. Observations:                 392   AIC:                             2063.
Df Residuals:                     384   BIC:                             2095.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const          -17.2184      4.644     -3.707   

i: displacements, weight, year, and origin 

ii: cylinders, horsepower, acceleration

iii: one unit uncrease in year is associated with 0.75 unit increase in mpg everything else being equal

### e)

cylinders, acceleration, and origin look most likely to have a nonlinear relationship with mpg.

In [5]:
import numpy as np
df_quant['cylinders_sqr']=np.square(df_quant['cylinders'])
df_quant['displacement_sqr']=np.square(df_quant['displacement'])
df_quant['origin_sqr']=np.square(df_quant['origin'])
df_quant['acceleration_sqr']=np.square(df_quant['acceleration'])
x2 = sm.add_constant(df_quant[['cylinders','displacement','horsepower','weight',
                              'acceleration','year','origin',
                               'cylinders_sqr','displacement_sqr','origin_sqr','acceleration_sqr']])
reg2 = sm.OLS(y, x2)
results2 = reg2.fit()
print(results2.summary())

                            OLS Regression Results                            
Dep. Variable:                    mpg   R-squared:                       0.858
Model:                            OLS   Adj. R-squared:                  0.853
Method:                 Least Squares   F-statistic:                     207.9
Date:                Tue, 18 Feb 2020   Prob (F-statistic):          2.99e-153
Time:                        20:39:44   Log-Likelihood:                -979.27
No. Observations:                 392   AIC:                             1983.
Df Residuals:                     380   BIC:                             2030.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -0.2019      7.474  

i: Adjusted $R^2$ is 0.853 which is larger than 0.818 in part (d). This means the model in part(e) better fits the data.

ii: The coefficients on displacement and squared displacement are both statistically significant at the 1% level.

iii: The coefficients on cylinders and squared cylinders are not statistically significant at the 10% level.

### f)

In [6]:
results2.predict([1, 6, 200, 100, 3100, 15.1, 99, 1, 6**2, 200**2, 1, 15.1**2])

array([38.67266199])

## 2

### a)

In [7]:
X1 = [0, 2, 0, 0, -1, 1]
X2 = [3, 0, 1, 1, 0, 1]
X3 = [0, 0, 3, 2, 1, 1]
Y = ['Red', 'Red', 'Red', 'Green', 'Green', 'Red']
knn = pd.DataFrame(data={'X1':X1, 'X2':X2, 'X3':X3, 'Y':Y})
knn['distance'] = round(np.sqrt(knn.X1**2 + knn.X2**2 + knn.X3**2),2)
knn

X1  X2  X3      Y  distance
0   0   3   0    Red      3.00
1   2   0   0    Red      2.00
2   0   1   3    Red      3.16
3   0   1   2  Green      2.24
4  -1   0   1  Green      1.41
5   1   1   1    Red      1.73

### b)

For K = 1, the nearest one is the 5th observation with a distance 1.41. And the 5th observation is Green. So our prediction is green.

### c)

For K = 3, the nearest 3 observations are the 2nd, 5th and 6th observation.Two of them are red and one of them is green. So our prediction is red.

### d)

If the Bayes (optimal) decision boundary in this problem is highly non-linear, this classifier is likely overfitting the data so we expect N to be small.

### e)

In [8]:
from sklearn.neighbors import KNeighborsClassifier
knn_p = KNeighborsClassifier(n_neighbors = 2)
predictors = list(zip(knn['X1'], knn['X2'], knn['X3']))
knn_p.fit(predictors, knn['Y'])
prediction = knn_p.predict([[1, 1, 1]])[0]
print("The KNN classifier prediction:", prediction)

The KNN classifier prediction: Green


### 3)

In [9]:
df_quant['mpg high'] = np.where(df_quant['mpg']>np.median(df_quant['mpg']),1,0)

In [10]:
log_x = sm.add_constant(df_quant[['cylinders','displacement','horsepower','weight',
                              'acceleration','year','origin']])
log_y = df_quant['mpg high']
log_reg1 = sm.Logit(log_y, log_x, missing = 'drop')
log_results1 = log_reg1.fit()

Optimization terminated successfully.
         Current function value: 0.200944
         Iterations 9


In [11]:
print(log_results1.summary2())

                         Results: Logit
Model:              Logit            Pseudo R-squared: 0.710     
Dependent Variable: mpg high         AIC:              173.5398  
Date:               2020-02-18 20:39 BIC:              205.3099  
No. Observations:   392              Log-Likelihood:   -78.770   
Df Model:           7                LL-Null:          -271.71   
Df Residuals:       384              LLR p-value:      2.5305e-79
Converged:          1.0000           Scale:            1.0000    
No. Iterations:     9.0000                                       
-----------------------------------------------------------------
                 Coef.   Std.Err.    z    P>|z|   [0.025   0.975]
-----------------------------------------------------------------
const           -17.1549   5.7638 -2.9763 0.0029 -28.4517 -5.8580
cylinders        -0.1626   0.4232 -0.3842 0.7008  -0.9920  0.6669
displacement      0.0021   0.0120  0.1741 0.8618  -0.0215  0.0257
horsepower       -0.0410   0.0239 -1

The regressors that have coefficients that are statistically significant at the 5% level:
weight and year 

### b)

In [12]:
import sklearn
from sklearn.model_selection import train_test_split

In [13]:
X_train, X_test, y_train, y_test = train_test_split(log_x, log_y,
                                                    test_size = 0.5, random_state=10)

### c)

In [14]:
from sklearn.linear_model import LogisticRegression
log2_results2 = LogisticRegression().fit(X_train, y_train)
coefficients = log2_results2.coef_
names = log_x.columns.tolist()
coefficients_list = coefficients.tolist()[0]
df_logit = pd.DataFrame(data={'Regressors':names, 'coefficients':coefficients_list})
df_logit

Regressors  coefficients
0         const     -0.070226
1     cylinders     -0.676048
2  displacement      0.006087
3    horsepower     -0.038023
4        weight     -0.005055
5  acceleration     -0.134894
6          year      0.299868
7        origin     -0.154037

In [15]:
y_pred = log2_results2.predict(X_test)

In [16]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_pred)
confusion_matrix

array([[86, 13],
       [12, 85]], dtype=int64)

In [18]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

             precision    recall  f1-score   support

          0       0.88      0.87      0.87        99
          1       0.87      0.88      0.87        97

avg / total       0.87      0.87      0.87       196



F1 Score is the weighted average of Precision and Recall. Therefore, this score takes both false positives and false negatives into account. We have same f1-score for 0 and 1. So this model predicts equally well on low mpg and high mpg.